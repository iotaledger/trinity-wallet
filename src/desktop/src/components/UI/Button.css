@keyframes spin {
    to {
        transform: rotate(360deg);
    }
}

.button {
    position: relative;
    border: 1px solid transparent;
    border-radius: 3px;
    cursor: pointer;
    display: inline-block;
    height: 42px;
    line-height: 40px;
    font-size: var(--font-size-s);
    letter-spacing: 0.2px;
    min-width: 140px;
    text-decoration: none;
    text-transform: uppercase;
    font-weight: 200;
    padding: 0 24px 0;
    color: var(--color-Border);
    user-select: none;
    overflow: visible;
    transition: padding 0.2s cubic-bezier(0.55, 0, 0.1, 1);
    &:focus {
        outline: none;
    }
    &.small {
        padding: 0 16px;
        height: 32px;
        line-height: 32px;
        min-width: 30px;
        border-radius: 2px;
    }
    &:disabled,
    &.disabled {
        opacity: 0.5;
        pointer-events: none;
    }
    &:before {
        content: '';
        box-sizing: border-box;
        position: absolute;
        top: 50%;
        left: 18px;
        width: 16px;
        height: 16px;
        margin: -8px 0 0 -8px;
        border-radius: 50%;
        border: 1px solid transparent;
        border-top-color: var(--color-Border);
        opacity: 0;
    }
    &:after {
        content: '';
        position: absolute;
        width: calc(100% + 2px);
        height: calc(100% + 2px);
        border-radius: 3px;
        top: -1px;
        left: -1px;
        box-shadow: inset 0 0 0 30px var(--hover);
        opacity: 0;
    }
    &:hover:after {
        opacity: 0.8;
    }
    &:active:after {
        opacity: 1;
    }
}

footer .button + .button {
    margin-left: 20px;
}

.primary {
    composes: button;
    background: var(--positive-bg);
    border-color: var(--positive-border);
}

.secondary {
    composes: button;
    background: var(--secondary-bg);
    border-color: var(--secondary-border);
}

.warning {
    composes: button;
    background: var(--negative-bg);
    border-color: var(--negative-border);
}

.highlight {
    composes: button;
    background: var(--highlight-bg);
    border-color: var(--highlight-border);
}

.extra {
    composes: button;
    background: var(--extra-bg);
    border-color: var(--extra-border);
}

.outline {
    background: none;
    border: 1px solid;

    &.primary {
        color: var(--cta);
        border-color: var(--cta);
    }

    &.secondary {
        color: var(--secondary);
        border-color: var(--secondary);
    }
    &.warning {
        color: var(--negative);
        border-color: var(--negative);
    }
    &.extra {
        color: var(--extra);
        border-color: var(--extra);
    }
    &.highlight {
        color: var(--highlight);
        border-color: var(--highlight);
    }
}

.loading {
    pointer-events: none;
    padding: 0 12px 0 36px;
    box-shadow: inset 0 0 0 30px var(--color-hover);
    &:before {
        transition: all 0.8s cubic-bezier(0.55, 0, 0.1, 1);
        opacity: 1;
        animation: spin 0.6s ease infinite;
    }
}
