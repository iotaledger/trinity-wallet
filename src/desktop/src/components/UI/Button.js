import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Link } from 'react-router-dom';
import css from './Button.css';

export default class Button extends React.PureComponent {
    static propTypes = {
        children: PropTypes.node,
        className: PropTypes.string,
        loading: PropTypes.bool,
        variant: PropTypes.oneOf(['primary', 'secondary', 'success', 'warning', 'danger', 'info', 'extra']).isRequired,
        to: PropTypes.string,
    };

    static defaultProps = {
        variant: 'primary',
    };

    render() {
        const { children, className, to, variant, loading } = this.props;

        if (to) {
            return (
                <Link {...this.props} className={classNames(css[className], css[variant])}>
                    {children}
                </Link>
            );
        }

        return (
            <button {...this.props} className={classNames(css[className], css[variant], loading ? css.loading : null)}>
                {children}
            </button>
        );
    }
}
