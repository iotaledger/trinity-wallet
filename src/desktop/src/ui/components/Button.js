import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Link } from 'react-router-dom';
import css from './button.scss';

/**
 * Button component
 */
export default class Button extends React.PureComponent {
    static propTypes = {
        /** Target link */
        to: PropTypes.string,
        /** Click event callback function */
        /** @param {object} event - Click event object */
        onClick: PropTypes.func,
        /** Button or link element content */
        children: PropTypes.node,
        /** Custom button style definitions */
        style: PropTypes.object,
        /** Button disabled state */
        disabled: PropTypes.bool,
        /** Button type */
        type: PropTypes.string,
        /** Buttons style type */
        variant: PropTypes.oneOf(['primary', 'secondary', 'positive', 'negative', 'extra', 'back']).isRequired,
        /** Buttons custom class */
        className: PropTypes.oneOf(['outline', 'small', 'large', 'square', 'inline', 'icon']),
        /** Buttons loading state */
        loading: PropTypes.bool,
    };

    static defaultProps = {
        variant: 'primary',
    };

    render() {
        const { onClick, children, className, to, variant, loading, style, type, disabled } = this.props;

        const loadingClass = loading ? css.loading : null;

        if (to) {
            return (
                <Link {...this.props} className={classNames(css.button, css[className], css[variant])}>
                    {children}
                </Link>
            );
        }

        return (
            <button
                style={style}
                type={type ? type : 'button'}
                onClick={onClick}
                disabled={disabled}
                className={classNames(css.button, css[className], css[variant], loadingClass)}
            >
                {children}
            </button>
        );
    }
}
