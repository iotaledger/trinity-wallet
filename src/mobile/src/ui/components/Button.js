import React, { PureComponent } from 'react';
import { View, Text, StyleSheet, TouchableOpacity } from 'react-native';
import PropTypes from 'prop-types';
import { width, height } from 'mobile/src/libs/dimensions';
import General from 'mobile/src/ui/theme/general';

const styles = StyleSheet.create({
    container: {
        alignItems: 'center',
    },
    children: {
        fontFamily: 'SourceSansPro-SemiBold',
        fontSize: General.fontSize3,
        textAlign: 'center',
    },
    wrapper: {
        justifyContent: 'center',
        alignItems: 'center',
        width,
        height: height / 11,
    },
});

export default class Button extends PureComponent {
    static propTypes = {
        /** Press event callback function */
        onPress: PropTypes.func.isRequired,
        /** Button text */
        children: PropTypes.string.isRequired,
        /** Component styles to override default ones */
        style: PropTypes.object,
        /** Id for automated screenshots */
        testID: PropTypes.string,
    };

    static defaultProps = {
        style: {},
        testID: '',
    };

    render() {
        const { style, children, testID } = this.props;

        return (
            <View style={[styles.container, style.container]}>
                <TouchableOpacity onPress={() => this.props.onPress()} testID={testID}>
                    <View style={[styles.wrapper, style.wrapper]}>
                        <Text style={[styles.children, style.children]}>{children}</Text>
                    </View>
                </TouchableOpacity>
            </View>
        );
    }
}
