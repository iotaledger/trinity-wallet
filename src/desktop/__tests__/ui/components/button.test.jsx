import React from 'react';
import { shallow } from 'enzyme';

import Button from 'ui/components/Button';

const props = {
    onClick: jest.fn(),
    variant: 'primary',
};

describe('Button component', () => {
    test('Render the component', () => {
        const wrapper = shallow(<Button {...props} />);

        expect(wrapper).toMatchSnapshot();
    });

    test('Button click event', () => {
        const wrapper = shallow(<Button {...props} />);

        wrapper.simulate('click');
        expect(props.onClick).toHaveBeenCalledTimes(1);
    });

    test('Button variant class name', () => {
        const wrapper = shallow(<Button {...props} />);

        expect(wrapper.hasClass('primary')).toBeTruthy();
    });

    test('Button additional class name', () => {
        const mockProps = Object.assign({}, props, { className: 'outline' });
        const wrapper = shallow(<Button {...mockProps} />);

        expect(wrapper.hasClass('outline')).toBeTruthy();
    });

    test('Button loading state', () => {
        const mockProps = Object.assign({}, props, { loading: true });
        const wrapper = shallow(<Button {...mockProps} />);

        expect(wrapper.hasClass('loading')).toBeTruthy();
    });

    test('Button disabled state', () => {
        const mockProps = Object.assign({}, props, { disabled: true });
        const wrapper = shallow(<Button {...mockProps} />);

        expect(wrapper.props().disabled).toBeTruthy();
    });
});
