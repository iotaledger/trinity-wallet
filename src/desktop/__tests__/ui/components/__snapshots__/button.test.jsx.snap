// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Render the component 1`] = `
<button
  className="button undefined primary"
  onClick={[MockFunction]}
  type="button"
/>
`;
